### PR TITLE
Moves to 'grid' terminology

### PR DESCRIPTION
**Pull request by gkanwar**
'lattice' -> 'grid'
'link' -> 'edge'

Also reorders the grid dimensions argument in the typedef parsing, such that it should now be grid{<edge type>}[<ndims>](<underlying>).

The term "Unstructured" is still used in the code itself to contrast with "Grid". We may want to simplify this but I wanted to get the frontend change out quickly, so I will address that in a future PR.

### CODE REVIEW

File: test/input/illegal/declarations.sim [hunk 1]
```diff
@@ -5,21 +5,21 @@ end
...
 
-%%% bad-lattice-link-set-type-1
+%%% bad-grid-link-set-type-1
 element E
 end
 
 extern Q : set{E};
 extern P : set{E}(Q);
-extern S : lattice[1]{E}(P);
+extern S : grid{E}[1](P);
```
Comment (fredrikbk) [inline test/input/illegal/declarations.sim]: I'm really sorry about the back and forth(!!), but now I'm thinking maybe your original `grid[1]{E}(P)` is better, since the dimension refers to the `grid` and both `E` and `P`. What do you think?

Depending on whether you think this this makes more sense, I'm happy to make the change if you'd like since I'm the one causing the back and forth.

File: src/types.h [hunk 2]
```diff
@@ -262,19 +262,19 @@ struct UnstructuredSetType : SetType {
...
-  /// determination of boundary conditions is also delegated to the lattice link
+  /// determination of boundary conditions is also delegated to the grid edge
   /// set definition, though we assume periodic for now.
   size_t dimensions;
-  /// Underlying point set of the lattice. Elements of this edge set connect
-  /// neighboring grid points in the lattice.
-  IndexSet latticePointSet;
+  /// Underlying point set of the grid. Elements of this edge set connect
+  /// neighboring grid points in the grid.
+  IndexSet underlyingPointSet;
```
Comment (fredrikbk) [inline src/types.h]: How about just `gridPoints`?

**Comment by fredrikbk** [comment]
I had a couple of comments, one related to the syntax. Otherwise it looks good and I'll merge it once we've decided on the syntax.

File: test/input/illegal/declarations.sim (hunk 1 above)
Comment (gkanwar) [inline test/input/illegal/declarations.sim]: Reverted to the previous format, since I agree that seems to match the semantics a little better.

File: src/types.h (hunk 2 above)
Comment (gkanwar) [inline src/types.h]: I was hoping to make it absolutely clear that this set is defining an edge set _over_ a set of points, as opposed to the points themselves, so I would advocate for leaving this as-is. If it's too wordy, perhaps `underlyingPoints` would be better?

**Comment by fredrikbk** [comment]
Looks like this branch has some issues with the llvm 3.7 merge

**Comment by gkanwar** [comment]
Rebased onto the latest master, this should be okay to merge now.
